mdpow/analysis/ensemble.py

```diff
@@ -0,0 +1,525 @@
+
+    The class is designed based on the `AnalysisBase
+    class in MDAnalysis <https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html>`_
+    and is a template for creating multi-universe multi-frame analyses using the
+    :class:`~mdpow.analysis.ensemble.Ensemble` object
+
+    :Keywords:
+    *ensemble*
+        The :class:`~mdpow.analysis.ensemble.Ensemble` being analyzed in the class
+
+
+    .. rubric:: Example Analysis
+
+    Dihedral Analysis Demonstration::
+
+        class DihedralAnalysis(mdpow._ensemble.EnsembleAnalysis):
+            def __init__(self, DihedralEnsembleGroup):
+                super(DihedralAnalysis, self).__init__(DihedralEnsembleGroup._ensemble())
+
+                self._sel = DihedralEnsembleGroup
+
+            def _prepare_ensemble(self):
+                self.result_dict = {}
+                for s in ['water', 'octanol']:
+                    self.result_dict[s] = {'Coulomb': {},
+                                           'VDW': {}}
+                for key in self._sel.group_keys():
+                    self.result_dict[key[0]][key[1]][key[2]] = None
+
+            def _prepare_universe(self):
+                self.angle_dict = {'angle': None,
+                                   'time': None}
+                self.angles = []
+
+            def _single_frame(self):
+                angle = calc_dihedrals(self._sel[self._key].positions[0], self._sel[self._key].positions[1],
+                                       self._sel[self._key].positions[2], self._sel[self._key].positions[3])
+                self.angles.append(angle)
+
+            def _conclude_universe(self):
+                self.angle_dict['time'] = self.times
+                self.angle_dict['angle'] = self.angles
+                self.result_dict[self._key[0]][self._key[1]][self._key[2]] = self.angle_dict
+
+            def _conclude_ensemble(self):
+                self.results = pd.DataFrame(data=self.result_dict)
+
+        D = DihedralAnalysis.run(start=0 stop=10, step=1)
+
+    """
+
+    def __init__(self, ensemble=None):
+        self._ensemble = ensemble
+
+    def _setup_system(self, key, start=None, stop=None, step=None):
+        self._system = self._ensemble[key]
+        self._key = key
+        self.start = start
+        self.stop = stop
+        self.step = step
+        self._setup_frames(self._system.trajectory)
+
+    def _setup_frames(self, trajectory):
+        self._trajectory = trajectory
+        start, stop, step = trajectory.check_slice_indices(self.start, self.stop, self.step)
+        self.n_frames = len(range(start, stop, step))
+        self.frames = np.zeros(self.n_frames, dtype=int)
+        self.times = np.zeros(self.n_frames)
+
+    def _single_universe(self):
+        """Calculations on a single Universe object.
+
+            Run on each universe in the _ensemble during when
+            self.run in called.
+        """
+        pass  # pragma: no cover
```

Comment:
how is `_single_universe` different from `_prepare_universe`? Do you need both?

Reply:
`_prepare_universe` is setting up frame iteration, single universe is if the user wants to do some kind of method that runs on a full universe.